Test model

robot269006 opened the pull request:
# WHAT
Test message.rb model with rspec

# WHY
To check if the app is configured as I programmed it to be

robot269006 committed:
Install gem

1.Add gem 'rspec', gem 'faker', gem 'factorygirl'
2.bundle install
3.rails g rspec:install
4.Test with bundle exec spec

robot269006 committed:
Folder and file settings

1. Create ‘model’, ’controller’, ‘factory’ folders
2. Create message_rspec.rb -> under model folder
3. Create messages.rb files -> under factory folder

robot269006 committed:
Enable factorygirl

1. Add notion to rails_helper.rb

robot269006 committed:
Model test (factory-girl settings)

1. Factorygirl settings

robot269006 committed:
Model test (validity test)

1. Check bundle exec spec
   *no example found -> file name from .rspec to .spec
2. Check if the message is valid with body etc
3. Add validity test to when body is blank

robot269006 committed:
Model test (faker)

1. Faker description to body, created and updated dates

robot269006 commented:
<img width="501" alt="2017-05-06 14 43 51" src="https://cloud.githubusercontent.com/assets/24761206/25770214/7a0f325a-326a-11e7-86ea-fe0eb9fab9ea.png">

ninooo29 commented on spec/models/message_spec.rb:
```diff
@@ -0,0 +1,14 @@
+describe Message do
+  describe '#create' do
+    it 'is valid with body, user_id, group_id, created_at and updated_at' do
+      message = build(:message)
```
letを使ってmessageを定義するようにしましょう。

robot269006 replied:
ありがとうございます、修正させていただきました。

robot269006 committed:
Revision (let)

1. Use “let” to define “message”

ninooo29 commented on spec/models/message_spec.rb:
```diff
@@ -1,14 +1,18 @@
     end
+
     it 'is invalid with a blank body' do
       message = build(:message, body: "")
```
すいません、こちらもletで書きましょう。
修正していただいたletの使い方はあっています！

robot269006 replied:
ご確認ありがとうございます、対応させていただきました！

robot269006 committed:
Revision (let)

1. let used to set-up :message in the second test as well

hiroki5123 commented on spec/models/message_spec.rb:
```diff
@@ -0,0 +1,17 @@
+      message.valid?
+    end
+
+    let(:message) { build(:message, body: "" ) }
```
letでメッセージは生成されているので、it 'is invalid with a blank body' do ~ endのなかにbodyを空にする記述をしましょう。

robot269006 replied:
さきほどletにするようにとご指摘受けた気がするのですが、間違ってますでしょうか？
letはit do ~ endの中に書けないと思います（`let` is not available from within an exampleと出ます）が、元に戻せばいいのでしょうか？

yyuto6012 replied:
letはexampleの中では使えないので、valid以外のexampleの時はletで共通の処理をせず、そのexmpleの中でmessageをbuildして、特定の条件にあったmessageのインスタンスを使いましょう。という意味です！

robot269006 replied:
すみません、ちょっと内容が理解できてないのですが、修正した内容でも間違ってますでしょうか？invalidのexampleの中ではletを使わずに、message = build(:message, body: "")を中に書き込んでます。これででメッセージのインスタンスを作ってる作っていると思うのですが、もう少し具体的に修正内容を教えてください。よろしくお願いいたします。

yyuto6012 replied:
それで大丈夫です！

robot269006 commented on spec/models/message_spec.rb:
```diff
@@ -0,0 +1,17 @@
+      message.valid?
+    end
+
+    let(:message) { build(:message, body: "" ) }
```
承知しました（元の記述に戻します）、ご確認ありがとうございます！

robot269006 committed:
Revision (revert the second revision)

1. second invalidity test put back to using instance instead of “let”

yyuto6012 commented on spec/factories/messages.rb:
```diff
@@ -0,0 +1,9 @@
+FactoryGirl.define do
+  factory :message do
+    body       { Faker::Lorem.sentence }
+    user_id    1
```
user・groupは値を渡さなくても、
user
groupという形で自動でそれに基づくidが生成されます。

robot269006 replied:
ご確認ありがとうございます！少し入れ違いになってしまい失礼しました。

robot269006 committed:
Revision (factory-girl)

delete user&group_id -> it’s auto-generated anyway

yyuto6012 commented on spec/factories/messages.rb:
```diff
@@ -1,8 +1,6 @@
 FactoryGirl.define do
   factory :message do
     body       { Faker::Lorem.sentence }
-    user_id    1
```
user
groupっていう感じです。
user_id 1
groyo_id 1ではなく。

yyuto6012 replied:
修正して、テスト通る感じだったら、マージしてください。

robot269006 replied:
すみません、テスト通らないです。。。
以下のようになります：

user
group
→　テスト通りません

user_id
group_id
→　テスト通りません

user_id 1
group_id 1
→テスト通ります
　このパターンで書くかfakerで数字をランダムで出すかにした方がいいですか？

robot269006 committed:
Revert "Revision (factory-girl)"

This reverts commit 346c2f0d8482f52dd211babba00ae0275f4cc3de.

seoky0 commented:
テストが通らないのは、ユーザーやグループに対応するfactoryを定義していないからではないでしょうか？
一度確認をお願いいたします

robot269006 committed:
Revision (create users.rb and groups.rb factories)

1. Create groups.rb and users.rb as factories
2. Relate messages.rb to :user and :group when referring to the ids

seoky0 commented on spec/models/message_spec.rb:
```diff
@@ -0,0 +1,17 @@
+  describe '#create' do
+    let(:message) { build(:message) }
+    it 'is valid with body, user_id, group_id, created_at and updated_at' do
+      message.valid?
```
こちらマッチャが記述されていないですね

robot269006 replied:
失礼しました、追記しました

robot269006 committed:
Revision (matcher description)

added matcher description

seoky0 commented on spec/models/message_spec.rb:
```diff
@@ -0,0 +1,18 @@
+  describe '#create' do
+    let(:message) { build(:message) }
+    it 'is valid with body, user_id, group_id, created_at and updated_at' do
+      message.valid?
```
保存可能かの処理を確認する場合は、この記述は不要です

robot269006 replied:
ありがとうございます、修正しましたのでマージします。

robot269006 committed:
Revision (delete save check)

deleted message.valid? from check with filled message
